src/core/common/code_utils.hpp

```diff
@@ -122,4 +122,30 @@
         }                             \
     } while (false)
 
+#ifdef __cplusplus
+
+namespace ot {
+
+template <int> struct StaticAssertError;
+template <> struct StaticAssertError<true>
+{
+};
+
+} // namespace ot
+
+/**
+ * This macro does static assert.
+ *
+ * @param[in]   aCheck      The value to assert.
+ * @param[in]   aErrorType  Underscore connected error message type, e.g. invalid_heap_size.
+ *
+ */
+#define OT_STATIC_ASSERT(aCheck, aErrorType)                     \
+    {                                                            \
+        ot::StaticAssertError<(aCheck) != 0> ERROR_##aErrorType; \
```

Comment:
Nice template solution. :+1: 
Can we use a `typedef`? 

`typedef ot::StaticAssertError<(aCheck) != 0> ERROR_##aErrorType`

If the `typedef` approach works, then we avoid having a variable declaration. 
Also with `typedef` we may be able to use the `STATIC_ASSERT` in a class (definition) scope or global scope.

Reply:
Any reason to not use https://github.com/nestlabs/nlassert, which contains a portable static assertion facility that leverages C++11 and C11, when available?

Reply:
Tried out the `typedef`, it works but we get the warning "unused local type".

How about using a `enum` definition and `sizeof(type)`? 
(found this trick from boost library). 

```
#define OT_STATIC_ASSERT(aCheck, aErrorType)                     \
    enum { \
        ERROR_##aErrorType = sizeof(ot::StaticAssertError<(aCheck) != 0>), \
    }
```
I tried this and it works.  We don't seem to get the "unused value" warning for `enum` definition. 
We can also use this in global scope or inside class definition. 

Thoughts?

Reply:
@gerickson This assert is supposed to be used by core code only, which uses C++03 only. If a few  lines of code could do the job, I would prefer not involving a whole library. 

@abtink thanks for your suggestions, I will try your version today.

Reply:
It's a header-only library and it works with both C++03 and C++11, portably, without compiler errors or warnings. Ostensibly, OpenThread would like to use asserts elsewhere in the code.

From that perspective, it makes sense to use a well-tested and broadly-used assertion implementation for which optimizations and improvements benefit all users (Weave, etc.).

Reply:
@gerickson I studied the implementation of nlassert. As you mentioned, it leverages to native implementation when possible, which is very beautiful just like boost's implementation. However, OpenThread explicitly disables C++11, the leverage will actually not be used. If one day, we are going to support C++11, we can add the leverage then. Besides, nlassert also provides runtime assertions, which is duplicate to OpenThread's existing implementation.  Even we insist adding C++11 leverage now, we still uses only a few lines of code from nlassert. Utilities in code_utils.hpp are supposed and designed to be used by core code only. We'd like to keep it small, simple , style consistent and easy to understand, so that new developers can quickly get started to contribute. Users should choose their own favorite assert libraries, such as nlassert, boost or C++11. Thoughts?